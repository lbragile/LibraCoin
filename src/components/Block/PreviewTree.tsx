import React, { useRef } from "react";
import { Table } from "react-bootstrap";
import { useAppContext } from "../../hooks/useAppContext";
import { useDrawPreviewTree } from "../../hooks/useDrawPreviewTree";

import "./Block.scss";

export default function PreviewTree(): JSX.Element {
  const { state } = useAppContext();

  const numCells = useRef<number>(9);
  const strLen = useRef<number>(20);
  const tree = useDrawPreviewTree(state.verifiedTrans, state.selectedTrans);

  // https://stackoverflow.com/a/1199420/4298115
  function truncate(str: string, n = strLen.current): string {
    return str?.length > n ? str.substr(0, n - 1) + "..." : str;
  }

  function getText(rowNum: number, index: number, isTitle: boolean): string {
    let text = "";

    switch (rowNum) {
      case 0: {
        if (index === Math.floor(numCells.current / 2)) {
          text = isTitle ? tree[tree.length - 1][0] : truncate(tree[tree.length - 1][0]);
        }
        break;
      }

      case 1: {
        if (index === numCells.current / 3 - 1) {
          text = isTitle ? tree[tree.length - 2][0] : truncate(tree[tree.length - 2][0]);
        } else if (index === (numCells.current / 3) * 2) {
          text = isTitle ? tree[tree.length - 2][1] : truncate(tree[tree.length - 2][1]);
        }
        break;
      }

      case 2: {
        if (index === 0) {
          text = isTitle ? tree[0][0] : truncate(tree[0][0]);
        } else if (index === numCells.current / 3 - 1) {
          text = isTitle ? tree[0][1] : truncate(tree[0][1]);
        } else if (index === (numCells.current / 3) * 2) {
          text = isTitle ? tree[0][2] : truncate(tree[0][2]);
        } else if (index === numCells.current - 1) {
          text = isTitle ? tree[0][3] : truncate(tree[0][3]);
        }
        break;
      }

      // no default
    }

    return text;
  }

  function getClassName(rowNum: number, index: number): string {
    let className = "";

    switch (rowNum) {
      case 0:
        className = index === Math.floor(numCells.current / 2) ? "root-cell" : "";
        break;

      case 1: {
        if (index === numCells.current / 3) {
          className = "diag-line-left";
        } else if (index === (2 * numCells.current) / 3 - 1) {
          className = "diag-line-right";
        }

        break;
      }

      case 2: {
        const isCorrectColumn = [numCells.current / 3 - 1, (2 * numCells.current) / 3].includes(index);
        if (isCorrectColumn && tree.length <= 2) {
          className = "data-cells";
        } else if (isCorrectColumn) {
          className = "middle-cells";
        }

        break;
      }

      case 3: {
        if (index === 1) {
          className = "diag-line-left";
        } else if (index === numCells.current - 2 && tree[0].length === 4) {
          className = "diag-line-right";
        } else if ([numCells.current / 3 - 1, (numCells.current / 3) * 2].includes(index)) {
          className = "normal-line";
        }

        break;
      }

      case 4: {
        const isCorrectColumn = [0, numCells.current / 3 - 1, (numCells.current / 3) * 2].includes(index);
        if (isCorrectColumn || (numCells.current - 1 === index && tree[0].length === 4)) {
          className = "data-cells";
        }

        break;
      }

      // no default
    }

    return className;
  }

  return (
    <div className="w-100 my-3 text-center">
      <h3 className="font-weight-bold">Merkle Tree Visualization</h3>
      {tree[tree.length - 1][0] !== "" ? (
        <Table
          role="grid"
          aria-label="Preview Table"
          className="w-75 mx-auto my-1 text-center border"
          responsive
          size="sm"
        >
          <tbody role="presentation">
            <tr>
              {Array.from({ length: numCells.current }).map((_, i) => (
                <td key={"first-data-row-" + i} className={getClassName(0, i)} title={getText(0, i, true)}>
                  {getText(0, i, false)}
                </td>
              ))}
            </tr>
            {tree.length >= 2 && (
              <>
                <tr>
                  {Array.from({ length: numCells.current }).map((_, i) => (
                    <td key={"first-break-row-" + i} className={getClassName(1, i)} />
                  ))}
                </tr>
                <tr>
                  {Array.from({ length: numCells.current }).map((_, i) => (
                    <td key={"second-data-row-" + i} className={getClassName(2, i)} title={getText(1, i, true)}>
                      {getText(1, i, false)}
                    </td>
                  ))}
                </tr>
              </>
            )}
            {tree.length === 3 && (
              <>
                <tr>
                  {Array.from({ length: numCells.current }).map((_, i) => (
                    <td key={"second-break-row-" + i} className={getClassName(3, i)} />
                  ))}
                </tr>
                <tr>
                  {Array.from({ length: numCells.current }).map((_, i) => (
                    <td key={"third-data-row-" + i} className={getClassName(4, i)} title={getText(2, i, true)}>
                      {getText(2, i, false)}
                    </td>
                  ))}
                </tr>
              </>
            )}
          </tbody>
        </Table>
      ) : (
        <p className="my-3 border w-75 mx-auto">Please select a verified transaction from above...</p>
      )}
    </div>
  );
}
