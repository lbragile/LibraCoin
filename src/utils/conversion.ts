import { TextEncoder } from "text-encoding";

export function bufferToHex(hashBuffer: ArrayBuffer): string {
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(""); // convert bytes to hex string
  return hashHex;
}

export async function CryptoKeyToHex(format: string, key: CryptoKey): Promise<string> {
  const buf = (await window.crypto.subtle.exportKey(format, key)) as ArrayBuffer;
  return bufferToHex(buf);
}

export async function digestMessage(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8); // hash the message
  return bufferToHex(hashBuffer);
}

export function randomHash(len: number): string {
  return bufferToHex(window.crypto.getRandomValues(new Uint32Array(len)));
}
