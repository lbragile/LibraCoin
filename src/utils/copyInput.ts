export function copyKey(
  e: React.FocusEvent<HTMLTextAreaElement | HTMLInputElement>,
  setCopied: (arg: boolean[]) => void,
  type?: "public" | "private"
): void {
  e.target.select();
  e.target.setSelectionRange(0, 1e6);
  document.execCommand("copy");

  // wallet page, copying keys
  if (type) {
    const isPublic = type === "public";
    setCopied([isPublic, !isPublic && !e.target.value.includes("â—¦")]);
  } else {
    setCopied([true]);
  }
}
