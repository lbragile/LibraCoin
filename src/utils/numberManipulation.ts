// https://stackoverflow.com/a/2901298/4298115
export function numberWithCommas(x: number): string {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}
