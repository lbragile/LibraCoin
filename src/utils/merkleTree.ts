import { Tree } from "./Tree";
import { ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function calculateMerkleTreeFormation(
  verifiedTrans: ITransaction[],
  selectedTrans: ITransaction[],
  setMerkleTree?: (arg: string[][]) => void
): Promise<string> {
  let tree = [[""]];
  if (selectedTrans.length > 0) {
    // need to make sure node's in tree appear in same order as in the verified transaction pane, regardless of selection order
    const verifiedSignatures = verifiedTrans.map((trans) => trans.signature);
    const selectedSignatures = selectedTrans.map((trans) => trans.signature);
    let signatures = verifiedSignatures.filter((sig) => selectedSignatures.includes(sig));
    tree = [signatures];

    while (signatures.length !== 1) {
      const hashArr = [] as string[];
      for (let i = 0; i < signatures.length; i += 2) {
        const hash = signatures[i + 1] ? await digestMessage(signatures[i] + signatures[i + 1]) : signatures[i];
        hashArr.push(hash);
      }

      signatures = hashArr;
      tree.push(hashArr);
    }
  }

  if (setMerkleTree) {
    setMerkleTree(tree);
  }

  return getMerkleRoot(tree);
}

export function getMerkleRoot(tree: string[][]): string {
  return tree[tree.length - 1][0];
}

export function flattenTree(tree: string[][]): string[] {
  let flatTree = [] as string[];
  // need to reverse each tree row so that printing produces the right order!
  for (let i = 0; i < tree.length; i++) {
    flatTree = flatTree.concat(tree[i].reverse());
  }

  return flatTree;
}

export function drawTreeDiagramOnCanvas(
  merkleTree: string[][],
  canvas: HTMLCanvasElement | null,
  transactions: ITransaction[]
): void {
  if (canvas) {
    const canvasTree = new Tree(canvas, transactions);
    canvasTree.clear();
    const flatTree = flattenTree(merkleTree);
    for (let i = flatTree.length - 1; i >= 0; i--) {
      canvasTree.addNode(flatTree[i]);
    }
    canvasTree.drawTree();
  }
}
