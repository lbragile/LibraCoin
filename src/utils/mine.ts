import { digestMessage, randomHash } from "./conversion";

async function createTarget(numZeros: number): Promise<string> {
  const targetHash = await digestMessage(randomHash(20));

  // replace leading bits with zeros
  const re = new RegExp(`^.{0,${numZeros}}`, "g");
  const zerosStr = Array(numZeros).fill("0").join("");

  return targetHash.replace(re, zerosStr);
}

export async function mine(
  nonce: number,
  setHeader: (arg: number) => void,
  setTarget: (arg: string) => void,
  setSolution: (arg: string) => void,
  setIsValid: (arg: boolean) => void
): Promise<void> {
  // make target
  const numZeros = Math.round(Math.random()) + 2;
  const targetHash = await createTarget(numZeros);
  setTarget(targetHash);

  // mine
  let candidateSolution = "";
  let header = nonce; // re-assigning for clarity
  while (header <= Number.MAX_SAFE_INTEGER) {
    candidateSolution = await digestMessage(header.toString());
    setSolution(candidateSolution);
    setHeader(header++);

    const leadingBits = candidateSolution.substr(0, numZeros).split("");
    if (leadingBits.every((bit) => bit === "0")) {
      break;
    }
  }

  setIsValid(candidateSolution <= targetHash);
}
