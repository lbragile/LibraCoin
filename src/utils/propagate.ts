import { ACTIONS } from "../enums/AppDispatchActions";
import { IAction, IBlock, IState, ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function propagateBlockStatus(
  state: IState,
  dispatch: React.Dispatch<IAction>,
  index: number,
  prevHash: string,
  currHash: string,
  skipFirstUpdate: boolean,
  newRoot?: string,
  transactions?: ITransaction[]
): Promise<void> {
  const newBlocks: IBlock[] = [];
  for (let i = index; i < state.chain.length; i++) {
    const merkleRoot = newRoot && i === index ? newRoot : state.chain[i].merkleRoot;
    const valid = skipFirstUpdate && i === index;
    const showTrans = state.chain[i].showTrans ?? false;
    currHash = i === index ? currHash : await digestMessage(i + prevHash + merkleRoot);
    transactions = i === index && transactions ? transactions : state.chain[i].transactions;

    newBlocks.push({ index: i, timestamp: Date.now(), prevHash, currHash, transactions, merkleRoot, valid, showTrans });

    prevHash = currHash; // next block's prevHash is this block's currHash
  }

  dispatch({ type: ACTIONS.UPDATE_BLOCK, payload: { block: newBlocks } });
}
