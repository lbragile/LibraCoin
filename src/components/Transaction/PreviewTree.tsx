import React, { useContext, useEffect, useRef } from "react";
import { AppContext } from "../../context/AppContext";
import { IState } from "../../typings/AppTypes";
import { calculateMerkleTreeFormation, drawTreeDiagramOnCanvas } from "../../utils/merkleTree";
import { Tree } from "../../utils/Tree";

export default function PreviewTree(): JSX.Element {
  const { state } = useContext(AppContext) as { state: IState };

  const treeCanvas = useRef<HTMLCanvasElement | null>(null);

  useEffect(() => {
    async function drawTree() {
      if (treeCanvas.current) {
        const tree = await calculateMerkleTreeFormation(state.verifiedTrans, state.selectedTrans);
        const canvasTree = new Tree(treeCanvas.current, state.selectedTrans);
        drawTreeDiagramOnCanvas(tree, canvasTree);
      }
    }

    drawTree();
  }, [state.selectedTrans, state.verifiedTrans]);

  return (
    <div className="mb-2 d-none d-lg-block">
      <h4 className="font-weight-bold text-center">Merkle Tree Visualization</h4>
      <canvas ref={treeCanvas} className="border border-dark rounded" width={window.outerWidth * 0.7} />
    </div>
  );
}
