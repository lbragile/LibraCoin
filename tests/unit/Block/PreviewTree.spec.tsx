/**
 * @group unit
 */

import React from "react";
import { screen } from "@testing-library/react";

import * as ConversionUtil from "../../../src/utils/conversion";

import PreviewTree from "../../../src/components/Block/PreviewTree";
import { customRender } from "../../utils/testUtils";

const { initialState } = global;

it("renders correctly", async () => {
  const { asFragment } = customRender(<PreviewTree />);

  expect(await screen.findByRole("heading", { level: 3 })).toBeInTheDocument();
  expect(await screen.findByRole("grid", { name: /Preview Table/i })).toBeInTheDocument();
  expect(screen.queryByText("Please select a verified transaction from above...")).not.toBeInTheDocument();

  expect(asFragment()).toMatchSnapshot();
});

test("0 selected transactions", async () => {
  const { asFragment } = customRender(<PreviewTree />, { stateMock: { ...initialState, selectedTrans: [] } });

  expect(await screen.findByRole("heading", { level: 3 })).toBeInTheDocument();
  expect(screen.queryByRole("grid", { name: /Preview Table/i })).not.toBeInTheDocument();
  expect(await screen.findByText("Please select a verified transaction from above...")).toBeInTheDocument();

  expect(asFragment()).toMatchSnapshot();
});

test.each`
  num
  ${1}
  ${2}
  ${3}
  ${4}
`("$num selected transactions", async ({ num }) => {
  jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValue("random message");
  const selectedTrans = [...initialState.verifiedTrans.slice(0, num)];
  const { asFragment } = customRender(<PreviewTree />, { stateMock: { ...initialState, selectedTrans } });

  expect(await screen.findByRole("heading", { level: 3 })).toBeInTheDocument();
  expect(await screen.findByRole("grid", { name: /Preview Table/i })).toBeInTheDocument();
  expect(screen.queryByText("Please select a verified transaction from above...")).not.toBeInTheDocument();

  // when 3 transactions are selected, one of them is repeated in the preview at the second row
  selectedTrans.forEach((trans) => {
    const isRepeatedSig = num === 3 && trans.signature === selectedTrans[selectedTrans.length - 1].signature;
    expect(screen.getAllByTitle(trans.signature)).toHaveLength(Number(isRepeatedSig) + 1);
  });

  expect(asFragment()).toMatchSnapshot();
});
