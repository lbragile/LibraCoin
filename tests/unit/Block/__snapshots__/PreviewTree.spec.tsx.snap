// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 selected transactions 1`] = `
<DocumentFragment>
  <div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <p
      class="my-3 w-75 mx-auto text-light"
    >
      Please select a verified transaction from above...
    </p>
  </div>
</DocumentFragment>
`;

exports[`1 selected transactions 1`] = `
<DocumentFragment>
  .c0 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: transparent;
}

.c0.c0.c0 {
  border-top: none;
}

.c1 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2CA355;
}

.c1.c1.c1 {
  border-top: none;
}

<div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <div
      class="table-responsive"
    >
      <table
        aria-label="Preview Table"
        class="w-75 mx-auto my-1 text-center table table-sm"
        role="grid"
      >
        <tbody
          role="presentation"
        >
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c1"
              title="AB1.23"
              type="[object Object]"
            >
              AB1.23
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`2 selected transactions 1`] = `
<DocumentFragment>
  .c0 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: transparent;
}

.c0.c0.c0 {
  border-top: none;
}

.c1 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2CA355;
}

.c1.c1.c1 {
  border-top: none;
}

.c2 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top left,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c2.c2.c2 {
  border-top: none;
}

.c3 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top right,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c3.c3.c3 {
  border-top: none;
}

.c4 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2641A3;
}

.c4.c4.c4 {
  border-top: none;
}

<div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <div
      class="table-responsive"
    >
      <table
        aria-label="Preview Table"
        class="w-75 mx-auto my-1 text-center table table-sm"
        role="grid"
      >
        <tbody
          role="presentation"
        >
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c1"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c2"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c3"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="AB1.23"
              type="[object Object]"
            >
              AB1.23
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="AC456.78"
              type="[object Object]"
            >
              AC456.78
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`3 selected transactions 1`] = `
<DocumentFragment>
  .c0 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: transparent;
}

.c0.c0.c0 {
  border-top: none;
}

.c1 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2CA355;
}

.c1.c1.c1 {
  border-top: none;
}

.c2 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top left,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c2.c2.c2 {
  border-top: none;
}

.c3 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top right,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c3.c3.c3 {
  border-top: none;
}

.c4 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #EE964B;
}

.c4.c4.c4 {
  border-top: none;
}

.c5 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to left,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c5.c5.c5 {
  border-top: none;
}

.c6 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2641A3;
}

.c6.c6.c6 {
  border-top: none;
}

<div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <div
      class="table-responsive"
    >
      <table
        aria-label="Preview Table"
        class="w-75 mx-auto my-1 text-center table table-sm"
        role="grid"
      >
        <tbody
          role="presentation"
        >
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c1"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c2"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c3"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="AD999.99"
              type="[object Object]"
            >
              AD999.99
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c2"
              type="[object Object]"
            />
            <td
              class="c5"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c5"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c6"
              title="AB1.23"
              type="[object Object]"
            >
              AB1.23
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c6"
              title="AC456.78"
              type="[object Object]"
            >
              AC456.78
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c6"
              title="AD999.99"
              type="[object Object]"
            >
              AD999.99
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`4 selected transactions 1`] = `
<DocumentFragment>
  .c0 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: transparent;
}

.c0.c0.c0 {
  border-top: none;
}

.c1 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2CA355;
}

.c1.c1.c1 {
  border-top: none;
}

.c2 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top left,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c2.c2.c2 {
  border-top: none;
}

.c3 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to top right,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c3.c3.c3 {
  border-top: none;
}

.c4 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #EE964B;
}

.c4.c4.c4 {
  border-top: none;
}

.c5 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: linear-gradient(to left,transparent calc(50% - 1px),#777,transparent calc(50% + 1px));
}

.c5.c5.c5 {
  border-top: none;
}

.c6 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2641A3;
}

.c6.c6.c6 {
  border-top: none;
}

<div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <div
      class="table-responsive"
    >
      <table
        aria-label="Preview Table"
        class="w-75 mx-auto my-1 text-center table table-sm"
        role="grid"
      >
        <tbody
          role="presentation"
        >
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c1"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c2"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c3"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c4"
              title="random message"
              type="[object Object]"
            >
              random message
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c2"
              type="[object Object]"
            />
            <td
              class="c5"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
            <td
              class="c5"
              type="[object Object]"
            />
            <td
              class="c3"
              type="[object Object]"
            />
            <td
              class="c0"
              type="[object Object]"
            />
          </tr>
          <tr>
            <td
              class="c6"
              title="AB1.23"
              type="[object Object]"
            >
              AB1.23
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c6"
              title="AC456.78"
              type="[object Object]"
            >
              AC456.78
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c6"
              title="AD999.99"
              type="[object Object]"
            >
              AD999.99
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c6"
              title="AE987.65"
              type="[object Object]"
            >
              AE987.65
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: transparent;
}

.c0.c0.c0 {
  border-top: none;
}

.c1 {
  width: 11%;
  height: 35px;
  color: #FEFEFE;
  background: #2CA355;
}

.c1.c1.c1 {
  border-top: none;
}

<div
    class="w-100 mb-2 text-center"
  >
    <h3
      class="font-weight-bold w-25 mx-auto text-light"
    >
      Merkle Tree Visualization
    </h3>
    <div
      class="table-responsive"
    >
      <table
        aria-label="Preview Table"
        class="w-75 mx-auto my-1 text-center table table-sm"
        role="grid"
      >
        <tbody
          role="presentation"
        >
          <tr>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c1"
              title="AC456.78"
              type="[object Object]"
            >
              AC456.78
            </td>
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
            <td
              class="c0"
              title=""
              type="[object Object]"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
